Implement importmaps

Constellation opened the pull request:
#### 6c1996aea8b85babf563135acc2d05def8aca3cb
<pre>
Implement importmaps
<a href="https://bugs.webkit.org/show_bug.cgi?id=220823">https://bugs.webkit.org/show_bug.cgi?id=220823</a>
&lt;rdar://73478498&gt;

Reviewed by Youenn Fablet.

This patch implements import-maps[1], which integrates mapping between module specifier and actual URL so that we can import
modules with arbitrary module specifier. This helps integrating existing node.js / bun etc. modules into the web page without
bundlers.

1. JSC integrates importMapStatusPromise into its module loader pipeline. This makes us to wait for the importmap loading cleanly.
2. Fix a bug that &lt;script type=&quot;module&quot; src=&quot;URL&quot;&gt;&apos;s src is resolved. The spec does not resolve this URL and handle it as normal URL.
   The reason is that we would like to preload this resource so we need to have an URL here.
3. Fix WCIG import-mpas&apos; tests in WPT. These tests are indeterministic, so we modified WPT testharness.js and tests to make it deterministic.
4. For now, we workaround the iframe &apos;load&apos; event issue which blocks testing in WPT import-maps by modifying testharness.js. We should fix this
   issue in WebKit side later. They are located in LayoutTests/http/wpt/import-maps/data-driven.
   It should be removed once the issues are solved. <a href="https://bugs.webkit.org/show_bug.cgi?id=245772">https://bugs.webkit.org/show_bug.cgi?id=245772</a>
5. &quot;importmap&quot; &lt;script&gt; type and LoadableImportMap is integrated in WebCore. We share most of code with LoadableClassicScript by factoring out the
   common part as LoadableNonModuleScriptBase.
6. Core algorithm of ImportMap resolution is defined in JSC::ImportMap in runtime/ImportMap.h. This map is per JSGlobalObject, and implements the
   resolution algorithm. We would like to integrate this well in normal JSC framework in the future.
7. HTML script preloader is correctly configured to preload importmap as well.
8. Some of WPT tests are failing because of www1.localhost domain resolution issue.

[1]: <a href="https://github.com/WICG/import-maps">https://github.com/WICG/import-maps</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/import-meta/import-meta-resolve-importmap-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/specifier-error-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/import-maps/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/worker-request-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/worker-request.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/__dir__.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/cross-origin-bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/to-bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/to-data: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/unsafe-inline-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/unsafe-inline.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper-iframe.js: Added.
(event.catch):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper.js: Added.
(createTestIframe):
(resolve):
(async runTests):
(async export.async runTestsFromJSON):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/tools/format_json.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/tools/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/import-maps-base-url.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/import-maps-base-url.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/module-map-key-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/module-map-key.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/empty.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/inject-base.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/log.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/log.js.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/test-helper.js: Added.
(getHandlers):
(testInIframe):
(testScriptElement):
(testStaticImportInjectBase):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/script-supports-importmap-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/script-supports-importmap.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-import.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/w3c-import.log: Added.
* LayoutTests/tests-options.json:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::moduleLoaderImportModule):
(JSC::JSAPIGlobalObject::loadAndEvaluateJSScriptModule):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(visibility.PrivateRecursive.async loadModule):
(visibility.PrivateRecursive.async loadAndEvaluateModule):
(visibility.PrivateRecursive.async requestImportModule):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::moduleLoaderImportModule):
* Source/JavaScriptCore/parser/SourceProvider.h:
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::loadModule):
(JSC::importModule):
* Source/JavaScriptCore/runtime/Completion.h:
* Source/JavaScriptCore/runtime/ImportMap.cpp: Added.
(JSC::ImportMap::resolveImportMatch):
(JSC::parseURLLikeModuleSpecifier):
(JSC::ImportMap::resolve const):
(JSC::normalizeSpecifierKey):
(JSC::sortAndNormalizeSpecifierMap):
(JSC::ImportMap::registerImportMap):
* Source/JavaScriptCore/runtime/ImportMap.h: Copied from Source/WebCore/dom/ScriptElementCachedScriptFetcher.h.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::JSGlobalObject):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::isAcquiringImportMaps const):
(JSC::JSGlobalObject::setAcquiringImportMaps):
(JSC::JSGlobalObject::setPendingImportMaps):
(JSC::JSGlobalObject::clearPendingImportMaps):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::importMap const):
(JSC::JSGlobalObject::importMap):
(JSC::JSGlobalObject::importMapStatusPromise const):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSModuleLoader::loadModule):
(JSC::JSModuleLoader::requestImportModule):
* Source/JavaScriptCore/runtime/JSModuleLoader.h:
* Source/JavaScriptCore/runtime/LazyProperty.h:
(JSC::LazyProperty::isInitialized const):
* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::parseJSON):
* Source/WTF/wtf/JSONValues.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSExecState.h:
(WebCore::JSExecState::loadModule):
* Source/WebCore/bindings/js/ModuleFetchFailureKind.h:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::loadModuleScriptInWorld):
(WebCore::ScriptController::loadModuleScript):
(WebCore::ScriptController::setupModuleScriptHandlers):
(WebCore::ScriptController::reportExceptionFromScriptError):
(WebCore::ScriptController::registerImportMap):
(WebCore::ScriptController::isAcquiringImportMaps):
(WebCore::ScriptController::setAcquiringImportMaps):
(WebCore::ScriptController::setPendingImportMaps):
(WebCore::ScriptController::clearPendingImportMaps):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::resolveModuleSpecifier):
(WebCore::ScriptModuleLoader::resolve):
(WebCore::ScriptModuleLoader::responseURLFromRequestURL):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::createImportMetaProperties):
* Source/WebCore/dom/CurrentScriptIncrementer.h:
(WebCore::CurrentScriptIncrementer::CurrentScriptIncrementer):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::baseURLForComplete const):
(WebCore::Document::completeURL const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/InlineClassicScript.h:
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableClassicScript::takeError):
(WebCore::LoadableClassicScript::notifyFinished):
(WebCore::LoadableClassicScript::error const): Deleted.
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableImportMap.cpp: Copied from Source/WebCore/dom/LoadableClassicScript.cpp.
(WebCore::LoadableImportMap::create):
(WebCore::LoadableImportMap::LoadableImportMap):
(WebCore::LoadableImportMap::~LoadableImportMap):
(WebCore::LoadableImportMap::isLoaded const):
(WebCore::LoadableImportMap::takeError):
(WebCore::LoadableImportMap::wasCanceled const):
(WebCore::LoadableImportMap::notifyFinished):
(WebCore::LoadableImportMap::execute):
(WebCore::LoadableImportMap::load):
* Source/WebCore/dom/LoadableImportMap.h: Copied from Source/WebCore/dom/LoadableClassicScript.h.
(isType):
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::takeError):
(WebCore::LoadableModuleScript::error const): Deleted.
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.h:
* Source/WebCore/dom/PendingScript.cpp:
(WebCore::PendingScript::takeError):
(WebCore::PendingScript::error const): Deleted.
* Source/WebCore/dom/PendingScript.h:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ScriptElement):
(WebCore::ScriptElement::determineScriptType const):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
(WebCore::ScriptElement::executeScriptAndDispatchEvent):
(WebCore::ScriptElement::executePendingScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::scriptType const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::isClassicScript const):
(WebCore::ScriptElementCachedScriptFetcher::isModuleScript const):
(WebCore::ScriptElementCachedScriptFetcher::isImportMap const):
* Source/WebCore/dom/ScriptType.h: Copied from Source/WebCore/dom/InlineClassicScript.h.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(WebCore::CSSPreloadScanner::emitRule):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::createPreloadRequest):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
(WebCore::TokenPreloadScanner::StartTagScanner::shouldPreload):
* Source/WebCore/html/parser/HTMLResourcePreloader.cpp:
(WebCore::PreloadRequest::resourceRequest):
* Source/WebCore/html/parser/HTMLResourcePreloader.h:
(WebCore::PreloadRequest::PreloadRequest):
* Source/WebCore/html/parser/HTMLScriptRunner.cpp:
(WebCore::HTMLScriptRunner::runScript):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::loadModuleSynchronously):
(WebCore::WorkerOrWorkletScriptController::loadAndEvaluateModule):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::endElementNs):

Canonical link: <a href="https://commits.webkit.org/254987@main">https://commits.webkit.org/254987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d46bf9ab930da9dd214fc19b30bd380864695c25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90857 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100144 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33933 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83201 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96512 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/27035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77657 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26842 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81791 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81525 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69873 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35016 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15558 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32818 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16541 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3474 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36594 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39492 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38519 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35630 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/dom/LoadableClassicScript.cpp:
```diff
@@ -36,49 +36,44 @@
 
     if (m_cachedScript->errorOccurred())
-        return Error { ErrorType::CachedScript, std::nullopt, std::nullopt };
+        return Error { ErrorType::CachedScript, std::nullopt, { } };
```
Second std::null opt could also be { }

Constellation replied:
Changed.

youennf commented on Source/WebCore/dom/LoadableImportMap.cpp:
```diff
@@ -0,0 +1,74 @@
+    m_cachedScript = requestScriptWithCache(document, sourceURL, crossOriginMode(), String { integrity() }, priority());
+    if (!m_cachedScript)
+        return false;
+    m_cachedScript->addClient(*this);
```
It would read better if addClient was done in the same class as removeClient (so in LoadableNonModuleScriptBase).
We could probably also share some more code betweenthe two classes.

Constellation replied:
merged.

youennf commented on Source/WebCore/dom/LoadableScript.h:
```diff
@@ -38,12 +39,14 @@ class ScriptElement;
         Nosniff,
         FailedIntegrityCheck,
+        Resolve,
+        Script,
```
I am not very clear about the difference between ErrorType Script and CachedScript.

Constellation replied:
Rename CachedScript to Fetch.

youennf commented on Source/WebCore/dom/PendingScript.cpp:
```diff
@@ -79,9 +79,9 @@ bool PendingScript::isLoaded() const
 }
 
-bool PendingScript::error() const
+bool PendingScript::takeError()
```
Should we not return a std::optional<Error> given the name of the method.

Constellation replied:
Changed to hasError since it is only used for ASSERT.

youennf commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -370,6 +394,37 @@ bool ScriptElement::requestModuleScript(const TextPosition& scriptStartPosition)
+        return true;
+
+    callOnMainThread([this, element = Ref<Element>(m_element)] {
+        dispatchErrorEvent();
```
What ensures this to be valid?
Should we instead just capture element, and call element.dispatchEvent(...).
Also usually, when dispatching an event, we queue a task on the document. Can we do that here?

Constellation replied:
The behavior itself is valid since it is aligned to the spec https://html.spec.whatwg.org/multipage/scripting.html#prepare-a-script
> If src is the empty string, then [queue a task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-a-task) to [fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named [error](https://html.spec.whatwg.org/multipage/indices.html#event-error) at el, and return.

ScriptElement is inherited by multiple script element classes, and `dispatchErrorEvent` is a virtual function implemented by them, so we need to call that (otherwise the overridden function will be ignored). Default implementation of `dispatchErrorEvent` is dispatching an event to m_element. (SVGScriptElement has additional logic).

> Also usually, when dispatching an event, we queue a task on the document. Can we do that here?

This is nice! Changed it to use document's EventLoop's queueTask

youennf replied:
If we need 'this', should we capture a Ref then?

youennf commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -401,11 +456,48 @@ void ScriptElement::executeClassicScript(const ScriptSourceCode& sourceCode)
+
+    if (!m_isExternalScript) {
+        ASSERT(m_element.document().contentSecurityPolicy());
+        const ContentSecurityPolicy& contentSecurityPolicy = *m_element.document().contentSecurityPolicy();
```
const auto& seems to be used in this class for the same use.

Constellation replied:
Changed.

youennf commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -401,11 +456,48 @@ void ScriptElement::executeClassicScript(const ScriptSourceCode& sourceCode)
     // https://html.spec.whatwg.org/multipage/scripting.html#execute-the-script-block
 
-    ASSERT(!loadableModuleScript.error());
+    ASSERT(!loadableModuleScript.takeError());
```
It seems slightly strange for an ASSERT to take an error.

Constellation replied:
Changed it to `hasError()`.

youennf commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -433,43 +525,80 @@ void ScriptElement::dispatchLoadEventRespectingUserGestureIndicator()
-            if (auto* frame = m_element.document().frame())
-                frame->script().reportExceptionFromScriptError(error.value(), loadableScript.isModuleScript());
-        } else {
+    if (std::optional<LoadableScript::Error> error = loadableScript.takeError()) {
```
if (auto) ?

Constellation replied:
Changed.

youennf commented on Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
```diff
@@ -312,7 +312,7 @@ bool WorkerOrWorkletScriptController::loadModuleSynchronously(WorkerScriptFetche
                             LoadableScript::ErrorType::CachedScript,
                             std::nullopt,
-                            std::nullopt
+                            { }
```
s/std::nullopt/{ } for the first one?

Constellation replied:
Changed.

youennf reviewed (commented):
Since we are changing WPT tests, it would be better to land the WPT PR first.
Some nits below.

Constellation reviewed (commented):
Thanks!

MenloDorian commented on Source/JavaScriptCore/builtins/ModuleLoader.js:
```diff
@@ -378,15 +378,17 @@ function provideFetch(key, value)
     // Loader.resolve hook point.
     // resolve: moduleName => Promise(moduleKey)
     // Take the name and resolve it to the unique identifier for the resource location.
     // For example, take the "jquery" and return the URL for the resource.
```
This comment belongs in `loadAndEvaluateModule` and `requestImportModule` now.  OTOH, is this comment still valid even in those 2 functions?

Constellation replied:
I should just remove this comment, it is obsolete and just confusing.

MenloDorian commented on Source/JavaScriptCore/builtins/ModuleLoader.js:
```diff
@@ -409,15 +411,23 @@ async function loadAndEvaluateModule(moduleName, parameters, fetcher)
+    if (importMap)
+        await importMap;
+    var key = this.resolve(moduleName, @undefined, fetcher);
+    var key = await this.loadModule(key, parameters, fetcher);
```
Remove second `var` for `key`?

Constellation replied:
Fixed.

youennf reviewed (approved):
Let's move the tests requiring a WebKit workaround in LayoutTests/http/wpt.
And let's upstream the other changes.

Constellation committed:
Implement importmaps
https://bugs.webkit.org/show_bug.cgi?id=220823
<rdar://73478498>

Reviewed by Youenn Fablet.

This patch implements import-maps[1], which integrates mapping between module specifier and actual URL so that we can import
modules with arbitrary module specifier. This helps integrating existing node.js / bun etc. modules into the web page without
bundlers.

1. JSC integrates importMapStatusPromise into its module loader pipeline. This makes us to wait for the importmap loading cleanly.
2. Fix a bug that <script type="module" src="URL">'s src is resolved. The spec does not resolve this URL and handle it as normal URL.
   The reason is that we would like to preload this resource so we need to have an URL here.
3. Fix WCIG import-mpas' tests in WPT. These tests are indeterministic, so we modified WPT testharness.js and tests to make it deterministic.
4. For now, we workaround the iframe 'load' event issue which blocks testing in WPT import-maps by modifying testharness.js. We should fix this
   issue in WebKit side later. They are located in LayoutTests/http/wpt/import-maps/data-driven.
   It should be removed once the issues are solved. https://bugs.webkit.org/show_bug.cgi?id=245772
5. "importmap" <script> type and LoadableImportMap is integrated in WebCore. We share most of code with LoadableClassicScript by factoring out the
   common part as LoadableNonModuleScriptBase.
6. Core algorithm of ImportMap resolution is defined in JSC::ImportMap in runtime/ImportMap.h. This map is per JSGlobalObject, and implements the
   resolution algorithm. We would like to integrate this well in normal JSC framework in the future.
7. HTML script preloader is correctly configured to preload importmap as well.
8. Some of WPT tests are failing because of www1.localhost domain resolution issue.

[1]: https://github.com/WICG/import-maps

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/import-meta/import-meta-resolve-importmap-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/specifier-error-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/import-maps/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/dynamic-import.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload-link-header.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/modulepreload.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag-inline.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/script-tag.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/worker-request-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/acquiring/worker-request.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/__dir__.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/cross-origin-bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/to-bare: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/to-data: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/bare/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target-dynamic.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/applied-to-target.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash-failure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/hash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce-failure.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/nonce.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/unsafe-inline-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/unsafe-inline.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/csp/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resolving.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper-iframe.js: Added.
(event.catch):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper.js: Added.
(createTestIframe):
(resolve):
(async runTests):
(async export.async runTestsFromJSON):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/tools/format_json.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/tools/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-url-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/http-url-like-specifiers.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/import-maps-base-url.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/import-maps-base-url.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/module-map-key-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/module-map-key.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/multiple-import-maps/with-errors.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/not-as-classic-script.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/empty.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/inject-base.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/log.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/log.js.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/test-helper.js: Added.
(getHandlers):
(testInIframe):
(testScriptElement):
(testStaticImportInjectBase):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/script-supports-importmap-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/script-supports-importmap.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-import.py: Added.
(main):
* LayoutTests/imported/w3c/web-platform-tests/import-maps/w3c-import.log: Added.
* LayoutTests/tests-options.json:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::moduleLoaderImportModule):
(JSC::JSAPIGlobalObject::loadAndEvaluateJSScriptModule):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(visibility.PrivateRecursive.async loadModule):
(visibility.PrivateRecursive.async loadAndEvaluateModule):
(visibility.PrivateRecursive.async requestImportModule):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::moduleLoaderImportModule):
* Source/JavaScriptCore/parser/SourceProvider.h:
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::loadModule):
(JSC::importModule):
* Source/JavaScriptCore/runtime/Completion.h:
* Source/JavaScriptCore/runtime/ImportMap.cpp: Added.
(JSC::ImportMap::resolveImportMatch):
(JSC::parseURLLikeModuleSpecifier):
(JSC::ImportMap::resolve const):
(JSC::normalizeSpecifierKey):
(JSC::sortAndNormalizeSpecifierMap):
(JSC::ImportMap::registerImportMap):
* Source/JavaScriptCore/runtime/ImportMap.h: Copied from Source/WebCore/dom/ScriptElementCachedScriptFetcher.h.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::JSGlobalObject):
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::isAcquiringImportMaps const):
(JSC::JSGlobalObject::setAcquiringImportMaps):
(JSC::JSGlobalObject::setPendingImportMaps):
(JSC::JSGlobalObject::clearPendingImportMaps):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::importMap const):
(JSC::JSGlobalObject::importMap):
(JSC::JSGlobalObject::importMapStatusPromise const):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSModuleLoader::loadModule):
(JSC::JSModuleLoader::requestImportModule):
* Source/JavaScriptCore/runtime/JSModuleLoader.h:
* Source/JavaScriptCore/runtime/LazyProperty.h:
(JSC::LazyProperty::isInitialized const):
* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::parseJSON):
* Source/WTF/wtf/JSONValues.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSExecState.h:
(WebCore::JSExecState::loadModule):
* Source/WebCore/bindings/js/ModuleFetchFailureKind.h:
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::loadModuleScriptInWorld):
(WebCore::ScriptController::loadModuleScript):
(WebCore::ScriptController::setupModuleScriptHandlers):
(WebCore::ScriptController::reportExceptionFromScriptError):
(WebCore::ScriptController::registerImportMap):
(WebCore::ScriptController::isAcquiringImportMaps):
(WebCore::ScriptController::setAcquiringImportMaps):
(WebCore::ScriptController::setPendingImportMaps):
(WebCore::ScriptController::clearPendingImportMaps):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::resolveModuleSpecifier):
(WebCore::ScriptModuleLoader::resolve):
(WebCore::ScriptModuleLoader::responseURLFromRequestURL):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::createImportMetaProperties):
* Source/WebCore/dom/CurrentScriptIncrementer.h:
(WebCore::CurrentScriptIncrementer::CurrentScriptIncrementer):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::baseURLForComplete const):
(WebCore::Document::completeURL const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/InlineClassicScript.h:
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableClassicScript::takeError):
(WebCore::LoadableClassicScript::notifyFinished):
(WebCore::LoadableClassicScript::error const): Deleted.
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableImportMap.cpp: Copied from Source/WebCore/dom/LoadableClassicScript.cpp.
(WebCore::LoadableImportMap::create):
(WebCore::LoadableImportMap::LoadableImportMap):
(WebCore::LoadableImportMap::~LoadableImportMap):
(WebCore::LoadableImportMap::isLoaded const):
(WebCore::LoadableImportMap::takeError):
(WebCore::LoadableImportMap::wasCanceled const):
(WebCore::LoadableImportMap::notifyFinished):
(WebCore::LoadableImportMap::execute):
(WebCore::LoadableImportMap::load):
* Source/WebCore/dom/LoadableImportMap.h: Copied from Source/WebCore/dom/LoadableClassicScript.h.
(isType):
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::takeError):
(WebCore::LoadableModuleScript::error const): Deleted.
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.h:
* Source/WebCore/dom/PendingScript.cpp:
(WebCore::PendingScript::takeError):
(WebCore::PendingScript::error const): Deleted.
* Source/WebCore/dom/PendingScript.h:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ScriptElement):
(WebCore::ScriptElement::determineScriptType const):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
(WebCore::ScriptElement::executeScriptAndDispatchEvent):
(WebCore::ScriptElement::executePendingScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::scriptType const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::isClassicScript const):
(WebCore::ScriptElementCachedScriptFetcher::isModuleScript const):
(WebCore::ScriptElementCachedScriptFetcher::isImportMap const):
* Source/WebCore/dom/ScriptType.h: Copied from Source/WebCore/dom/InlineClassicScript.h.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(WebCore::CSSPreloadScanner::emitRule):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::createPreloadRequest):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
(WebCore::TokenPreloadScanner::StartTagScanner::shouldPreload):
* Source/WebCore/html/parser/HTMLResourcePreloader.cpp:
(WebCore::PreloadRequest::resourceRequest):
* Source/WebCore/html/parser/HTMLResourcePreloader.h:
(WebCore::PreloadRequest::PreloadRequest):
* Source/WebCore/html/parser/HTMLScriptRunner.cpp:
(WebCore::HTMLScriptRunner::runScript):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::loadModuleSynchronously):
(WebCore::WorkerOrWorkletScriptController::loadAndEvaluateModule):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::endElementNs):

Canonical link: https://commits.webkit.org/254987@main
